### PR TITLE
Decomp Heishi4 All OK

### PR DESCRIPTION
**Pull request by louist103**

### CODE REVIEW

**Review by Roman971 — CHANGES_REQUESTED** [review]
This file has some formatting issues that will be fixed with the format script. I'll wait for that before doing a full review so comments don't go outdated.

Also I believe this includes an accidental revert to an old version of the diff script.

File: src/overlays/actors/ovl_En_Heishi4/z_en_heishi4.c [hunk 1]
```diff
@@ -21,35 +33,352 @@ const ActorInit En_Heishi4_InitVars = {
     (ActorFunc)EnHeishi4_Update,
     (ActorFunc)EnHeishi4_Draw,
 };
-*/
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Heishi4/EnHeishi4_Init.s")
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Heishi4/EnHeishi4_Destroy.s")
+static u32 EnHeishi4ReactionSet[] = { 0x00000006, 0x00000007 };
```
Comment (Roman971) [inline src/overlays/actors/ovl_En_Heishi4/z_en_heishi4.c:37]:
```suggestion
static u32 sFaceReactionSets[] = { 6, 7 };
```

File: src/overlays/actors/ovl_En_Heishi4/z_en_heishi4.c [hunk 2]
```diff
@@ -21,35 +33,352 @@ const ActorInit En_Heishi4_InitVars = {
...
+            break;
+        case HEISHI4_MARKET_NIGHT_GUARD:
+            this->actionFunc = func_80A56544;
+            break;
+    }
+    this->unk_27C = ((thisx->params >> 8) & 0xFF);
+    osSyncPrintf("\n\n");
+    osSyncPrintf(VT_FGCOL(GREEN) " ☆☆☆☆☆ 兵士２セット完了！ ☆☆☆☆☆ %d\n" VT_RST, thisx->params);
+    osSyncPrintf(VT_FGCOL(YELLOW) " ☆☆☆☆☆ 識別完了！\t    ☆☆☆☆☆ %d\n" VT_RST, this->params);
+    osSyncPrintf(VT_FGCOL(PURPLE) " ☆☆☆☆☆ メッセージ完了！   ☆☆☆☆☆ %x\n\n" VT_RST, ((s32)thisx->params >> 8) & 0xF);
```
Comment (Roman971) [inline src/overlays/actors/ovl_En_Heishi4/z_en_heishi4.c:93]:
```suggestion
    osSyncPrintf(VT_FGCOL(PURPLE) " ☆☆☆☆☆ メッセージ完了！   ☆☆☆☆☆ %x\n\n" VT_RST, (thisx->params >> 8) & 0xF);
```

File: src/overlays/actors/ovl_En_Heishi4/z_en_heishi4.c [hunk 3]
```diff
@@ -21,35 +33,352 @@ const ActorInit En_Heishi4_InitVars = {
...
+    this->pos = thisx->posRot.pos;
+    thisx->unk_1F = 6;
+    if (this->params == 7) {
+        this->height = 30.0f;
+        ActorShape_Init(&thisx->shape, 0.0f, NULL, 30.0f);
+        SkelAnime_Init(globalCtx, &this->skelAnime, &D_0600BAC8, &D_0600C444, &this->limbDrawTable,
+                       &this->transitionDrawTable, 17);
+    } else {
+        this->height = 60.0f;
+        ActorShape_Init(&thisx->shape, 0.0f, &ActorShadow_DrawFunc_Circle, 30.0f);
```
Comment (Roman971) [inline src/overlays/actors/ovl_En_Heishi4/z_en_heishi4.c:66]:
```suggestion
        ActorShape_Init(&thisx->shape, 0.0f, ActorShadow_DrawFunc_Circle, 30.0f);
```

File: src/overlays/actors/ovl_En_Heishi4/z_en_heishi4.c [hunk 4]
```diff
@@ -21,35 +33,352 @@ const ActorInit En_Heishi4_InitVars = {
...
+    }
+    if (reactionOffset >= 3) {
+        reactionOffset = 1;
+    }
+    if (Text_GetFaceReaction(globalCtx, EnHeishi4ReactionSet[reactionOffset]) != 0) {
+        this->actor.textId = Text_GetFaceReaction(globalCtx, EnHeishi4ReactionSet[reactionOffset]);
+        this->unk_2B4 = 1;
+        this->actionFunc = func_80A56B40;
+    } else {
+        if ((gSaveContext.eventChkInf[8] & 1) != 0) {
```
Comment (Roman971) [inline src/overlays/actors/ovl_En_Heishi4/z_en_heishi4.c:126]:
```suggestion
        if (gSaveContext.eventChkInf[8] & 1) {
```

File: src/overlays/actors/ovl_En_Heishi4/z_en_heishi4.c [hunk 5]
```diff
@@ -21,35 +33,352 @@ const ActorInit En_Heishi4_InitVars = {
...
+            this->actor.textId = 0x5065;
+            this->actionFunc = func_80A56B40;
+            return;
+        }
+        if ((gSaveContext.eventChkInf[4] & 0x20) != 0) {
```
Comment (Roman971) [inline src/overlays/actors/ovl_En_Heishi4/z_en_heishi4.c:131]:
```suggestion
        if (gSaveContext.eventChkInf[4] & 0x20) {
```

File: src/overlays/actors/ovl_En_Heishi4/z_en_heishi4.c [hunk 6]
```diff
@@ -21,35 +33,352 @@ const ActorInit En_Heishi4_InitVars = {
...
+            return;
+        }
+        if (this->params == HEISHI4_IMPA) {
+            if (this->unk_284 == 0) {
+                this->actor.textId = 0x5079;
+            } else {
+                this->actor.textId = 0x507A;
+            }
+        } else {
+            if (!gSaveContext.nightFlag) {
```
Comment (Roman971) [inline src/overlays/actors/ovl_En_Heishi4/z_en_heishi4.c:143]: else if here imo

File: src/overlays/actors/ovl_En_Heishi4/z_en_heishi4.c [hunk 7]
```diff
@@ -21,35 +33,352 @@ const ActorInit En_Heishi4_InitVars = {
...
+                if (this->unk_284 == 0) {
+                    this->actor.textId = 0x5063;
+                } else {
+                    this->actor.textId = 0x5064;
+                }
+            } else {
+                if (this->unk_284 == 0) {
```
Comment (Roman971) [inline src/overlays/actors/ovl_En_Heishi4/z_en_heishi4.c:150]: else if here as well imo

File: src/overlays/actors/ovl_En_Heishi4/z_en_heishi4.c [hunk 8]
```diff
@@ -21,35 +33,352 @@ const ActorInit En_Heishi4_InitVars = {
...
+    if (reactionOffset >= 3) {
+        reactionOffset = 1;
+    }
+    if (Text_GetFaceReaction(globalCtx, EnHeishi4ReactionSet[reactionOffset]) != 0) {
+        this->actor.textId = Text_GetFaceReaction(globalCtx, EnHeishi4ReactionSet[reactionOffset]);
+        this->unk_2B4 = 1;
+        this->actionFunc = func_80A56B40;
+        return;
+    }
+    if (globalCtx->sceneNum == 77) {
```
Comment (Roman971) [inline src/overlays/actors/ovl_En_Heishi4/z_en_heishi4.c:190]: The scene enum value should be used here.

File: src/overlays/actors/ovl_En_Heishi4/z_en_heishi4.c [hunk 9]
```diff
@@ -21,35 +33,352 @@ const ActorInit En_Heishi4_InitVars = {
...
+        if (gSaveContext.nightFlag == 0) {
+            this->actor.textId = 0x7004;
+        } else {
+            this->actor.textId = 0x709A;
+        }
+    } else {
+        if (globalCtx->sceneNum != 33) {
```
Comment (Roman971) [inline src/overlays/actors/ovl_En_Heishi4/z_en_heishi4.c:197]: The scene enum value should be used here.
Also else if here imo

File: src/overlays/actors/ovl_En_Heishi4/z_en_heishi4.c [hunk 10]
```diff
@@ -21,35 +33,352 @@ const ActorInit En_Heishi4_InitVars = {
...
+        }
+    }
+    this->actionFunc = func_80A56B40;
+}
+
+void func_80A5673C(EnHeishi4* this, GlobalContext* globalCtx) {
+    if (gSaveContext.eventChkInf[4] & 0x20) {
+        osSyncPrintf(VT_FGCOL(YELLOW) " ☆☆☆☆☆ マスターソード祝入手！ ☆☆☆☆☆ \n" VT_RST);
+        Actor_Kill(&this->actor);
+    } else {
```
Comment (Roman971) [inline src/overlays/actors/ovl_En_Heishi4/z_en_heishi4.c:214]: early return here imo

File: src/overlays/actors/ovl_En_Heishi4/z_en_heishi4.c [hunk 11]
```diff
@@ -21,35 +33,352 @@ const ActorInit En_Heishi4_InitVars = {
...
+                this->actor.textId = 0x7007;
+                this->unk_282 = 5;
+                this->unk_284 = 1;
+                osSyncPrintf(VT_FGCOL(YELLOW) " ☆☆☆☆☆ デモ開始！ ☆☆☆☆☆ \n" VT_RST);
+            } else {
+                this->actor.textId = 0x7008;
+                this->unk_282 = 6;
+                osSyncPrintf(VT_FGCOL(BLUE) " ☆☆☆☆☆ 返事なし ☆☆☆☆☆ \n" VT_RST);
+            }
+            this->actionFunc = &func_80A56874;
```
Comment (Roman971) [inline src/overlays/actors/ovl_En_Heishi4/z_en_heishi4.c:229]:
```suggestion
            this->actionFunc = func_80A56874;
```

File: src/overlays/actors/ovl_En_Heishi4/z_en_heishi4.c [hunk 12]
```diff
@@ -21,35 +33,352 @@ const ActorInit En_Heishi4_InitVars = {
...
+            this->actionFunc = func_80A56A50;
+        }
+    }
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Heishi4/func_80A56614.s")
+void func_80A56A50(EnHeishi4* this, GlobalContext* globalCtx) {
+    f32 frames = SkelAnime_GetFrameCount(&D_0600C374);
+    this->unk_288 = frames;
+    SkelAnime_ChangeAnim(&this->skelAnime, &D_0600C374, 1.0f, 0.0f, frames, 2, -10.0f);
```
Comment (Roman971) [inline src/overlays/actors/ovl_En_Heishi4/z_en_heishi4.c:275]: Any chance the temp can be avoided here?
```suggestion
    this->unk_288 = SkelAnime_GetFrameCount(&D_0600C374);
    SkelAnime_ChangeAnim(&this->skelAnime, &D_0600C374, 1.0f, 0.0f, this->unk_288, 2, -10.0f);
```

File: src/overlays/actors/ovl_En_Heishi4/z_en_heishi4.c [hunk 13]
```diff
@@ -21,35 +33,352 @@ const ActorInit En_Heishi4_InitVars = {
...
+        }
+    }
+    if (func_8002F194(&this->actor, globalCtx) != 0) {
+        if ((this->params == HEISHI4_KAKRIKO_ENTRANCE) || (this->params == HEISHI4_IMPA)) {
+            this->unk_284 = 1;
+            this->actionFunc = func_80A563BC;
+            return;
+        }
+        if (this->params == HEISHI4_MARKET_NIGHT_GUARD) {
+            this->actionFunc = &func_80A56614;
```
Comment (Roman971) [inline src/overlays/actors/ovl_En_Heishi4/z_en_heishi4.c:330]:
```suggestion
            this->actionFunc = func_80A56614;
```

File: src/overlays/actors/ovl_En_Heishi4/z_en_heishi4.c [hunk 14]
```diff
@@ -21,35 +33,352 @@ const ActorInit En_Heishi4_InitVars = {
...
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Heishi4/func_80A56A50.s")
+void EnHeishi4_Update(Actor* thisx, GlobalContext* globalCtx) {
+    EnHeishi4* this = THIS;
+    s32 pad;
+    Player* player = PLAYER;
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Heishi4/func_80A56ACC.s")
+    // thisx->posRot.pos = this->pos;
```
Comment (Roman971) [inline src/overlays/actors/ovl_En_Heishi4/z_en_heishi4.c:342]:
```suggestion
```

File: src/overlays/actors/ovl_En_Heishi4/z_en_heishi4.c [hunk 15]
```diff
@@ -21,35 +33,352 @@ const ActorInit En_Heishi4_InitVars = {
...
+    return 0;
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Heishi4/func_80A56E14.s")
+void EnHeishi4_Draw(Actor* thisx, GlobalContext* globalCtx) {
+    EnHeishi4* this = THIS;
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Heishi4/EnHeishi4_Draw.s")
+    func_80093D18(globalCtx->state.gfxCtx);
+    SkelAnime_Draw(globalCtx, this->skelAnime.skeleton, this->skelAnime.limbDrawTbl, EnHeishi_OverrideLimbDraw, 0,
```
Comment (Roman971) [inline src/overlays/actors/ovl_En_Heishi4/z_en_heishi4.c:382]:
```suggestion
    SkelAnime_Draw(globalCtx, this->skelAnime.skeleton, this->skelAnime.limbDrawTbl, EnHeishi_OverrideLimbDraw, NULL,
```

File: src/overlays/actors/ovl_En_Heishi4/z_en_heishi4.h [hunk 16]
```diff
@@ -4,11 +4,37 @@
 #include <ultra64.h>
 #include <global.h>
 
+typedef enum {
+    /* 0x00 */ HEISHI4_KAKRIKO_ENTRANCE,
+    /* 0x04 */ HEISHI4_IMPA = 4,
+    /* 0x07 */ HEISHI4_DYING_GUARD = 7,
+    /* 0x08 */ HEISHI4_MARKET_NIGHT_GUARD,
+};
```
Comment (Roman971) [inline src/overlays/actors/ovl_En_Heishi4/z_en_heishi4.h:12]: This isn't necessarily the best but here are my suggestions to improve the consistency of this enum:
```suggestion
typedef enum {
    /* 0x00 */ HEISHI4_AT_KAKRIKO_ENTRANCE,
    /* 0x04 */ HEISHI4_AT_IMPAS_HOUSE = 4,
    /* 0x07 */ HEISHI4_AT_MARKET_NIGHT_DYING = 7,
    /* 0x08 */ HEISHI4_AT_MARKET_NIGHT
} Heishi4Type;
```
Also we usually have it below the struct but that's not an actual rule so I guess it's w/e

File: src/overlays/actors/ovl_En_Heishi4/z_en_heishi4.c [hunk 17]
```diff
@@ -21,35 +33,352 @@ const ActorInit En_Heishi4_InitVars = {
...
+extern AnimationHeader D_0600C444;
+extern AnimationHeader D_06005C30;
+extern AnimationHeader D_0600C6C8;
+extern AnimationHeader D_0600C374;
+
+void EnHeishi4_Init(Actor* thisx, GlobalContext* globalCtx) {
+    EnHeishi4* this = THIS;
+
+    Actor_SetScale(thisx, 0.01f);
+    this->params = thisx->params & 0xFF;
```
Comment (Roman971) [inline src/overlays/actors/ovl_En_Heishi4/z_en_heishi4.c:55]:
```suggestion
    this->type = thisx->params & 0xFF;
```

File: src/overlays/actors/ovl_En_Heishi4/z_en_heishi4.c [hunk 18]
```diff
@@ -21,35 +33,352 @@ const ActorInit En_Heishi4_InitVars = {
...
+    thisx->colChkInfo.mass = 0xFF;
+    this->pos = thisx->posRot.pos;
+    thisx->unk_1F = 6;
+    if (this->params == 7) {
```
Comment (Roman971) [inline src/overlays/actors/ovl_En_Heishi4/z_en_heishi4.c:59]: The type enum value should be used here

**Review by Roman971 — APPROVED** [review]

**Review by fig02 — APPROVED** [review]

**Pull request merged**